### ISSUE TITLE
NetJSON RFC JSON types

### ISSUE THREAD

**Issue by mitar** [interop-dev/netjson#46]
Not sure if this is the best place to comment on the RFC, but I see that for key names there are only primitive JSON types specified. Wouldn't it be better if a JSON schema type with JSON schema formatter definition would also be specified?

And along the lines of #45, it would be great if a semantic URI of the key definition would be specified/defined. It would be great to have semantic schema definitions for network objects in the standard.

**Comment by nemesisdesign** [comment]
For schema formatter definitions, do you mean things like:

```
{ "format": "ipv6" }
```

I'm currently working to implement this type of definition, I wrote down a note here:
https://github.com/openwisp/netjsonconfig/issues/42

In the last few months I have usually worked on the spec with a bottom up approach: I tried implementing things first, and then chose the best approach between simplicity and expressiveness.

While implementing things (especially in the [main netjsonconfig schema](https://github.com/openwisp/netjsonconfig/blob/master/netjsonconfig/schema.py)), and using existing implementations, I encountered a few issues that are pompting me to reconsider some initial ideas:

* I'm thinking about publishing a reference NetJSON schema implementation separately from the RFC; It would serve as a reference implementation for others to use as a base to build their own schema; this will allow the NetJSON-Schema reference implementation to evolve faster than the prose RFC, which once published should evolve at a slower rate
* formats in the current JSON-Schema v4 draft are not mandatory, so are not always implemented :-( - but I think it's a good idea to use them; we should also propose new formats like `subnet` (eg: 10.0.0.0/24) and `network` (eg: 192.168.5.24/23), but this will be a long term task

So :+1: for using formats (if I understood correctly) although I would prefer not to include the complete JSON-Schema in the RFC and just link to it.

Let me know what you think; I'd like to mark this as "accepted".

**Comment by mitar** [comment]
Yes, formatters is one (for strings). But maybe there are also other similar things. For example, for numeric values, we could define minimum and maximum values. Some values could be enums maybe. Things like that.
